File: src/kernel/syscalls/tunnel.cpp
#include "kernel/syscall_handlers.hpp"
#include "kernel/syscall_router.hpp"
#include "services/tunnel/client.hpp"
#include <spdlog/spdlog.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

namespace clove::kernel {

void TunnelSyscalls::register_syscalls(SyscallRouter& router) {
    router.register_handler(ipc::SyscallOp::SYS_TUNNEL_CONNECT,
        [this](const ipc::Message& msg) { return handle_tunnel_connect(msg); });
    router.register_handler(ipc::SyscallOp::SYS_TUNNEL_DISCONNECT,
        [this](const ipc::Message& msg) { return handle_tunnel_disconnect(msg); });
    router.register_handler(ipc::SyscallOp::SYS_TUNNEL_STATUS,
        [this](const ipc::Message& msg) { return handle_tunnel_status(msg); });
    router.register_handler(ipc::SyscallOp::SYS_TUNNEL_LIST_REMOTES,
        [this](const ipc::Message& msg) { return handle_tunnel_list_remotes(msg); });
    router.register_handler(ipc::SyscallOp::SYS_TUNNEL_CONFIG,
        [this](const ipc::Message& msg) { return handle_tunnel_config(msg); });
}

void TunnelSyscalls::on_tick() {
    process_tunnel_events();
}

ipc::Message TunnelSyscalls::handle_tunnel_connect(const ipc::Message& msg) {
    try {
        json j = json::parse(msg.payload_str());

        std::string relay_url = j.value("relay_url", context_.config.relay_url);
        std::string machine_id = j.value("machine_id", context_.config.machine_id);
        std::string token = j.value("token", context_.config.machine_token);

        if (relay_url.empty()) {
            json response;
            response["success"] = false;
            response["error"] = "relay_url required";
            return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONNECT, response.dump());
        }

        clove::services::tunnel::TunnelConfig tc;
        tc.relay_url = relay_url;
        tc.machine_id = machine_id;
        tc.token = token;
        context_.tunnel_client.configure(tc);

        if (context_.tunnel_client.connect()) {
            spdlog::info("Tunnel connected via syscall: {}", relay_url);
            json response;
            response["success"] = true;
            response["relay_url"] = relay_url;
            response["machine_id"] = machine_id;
            return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONNECT, response.dump());
        } else {
            json response;
            response["success"] = false;
            response["error"] = "Failed to connect to relay server";
            return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONNECT, response.dump());
        }

    } catch (const std::exception& e) {
        json response;
        response["success"] = false;
        response["error"] = std::string("invalid request: ") + e.what();
        return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONNECT, response.dump());
    }
}

ipc::Message TunnelSyscalls::handle_tunnel_disconnect(const ipc::Message& msg) {
    context_.tunnel_client.disconnect();

    json response;
    response["success"] = true;
    return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_DISCONNECT, response.dump());
}

ipc::Message TunnelSyscalls::handle_tunnel_status(const ipc::Message& msg) {
    auto status = context_.tunnel_client.get_status();

    json response;
    response["success"] = true;
    response["connected"] = status.connected;
    response["relay_url"] = status.relay_url;
    response["machine_id"] = status.machine_id;
    response["remote_agent_count"] = status.remote_agent_count;

    return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_STATUS, response.dump());
}

ipc::Message TunnelSyscalls::handle_tunnel_list_remotes(const ipc::Message& msg) {
    auto agents = context_.tunnel_client.list_remote_agents();

    json response;
    response["success"] = true;
    response["agents"] = json::array();

    for (const auto& agent : agents) {
        json a;
        a["agent_id"] = agent.agent_id;
        a["name"] = agent.name;
        a["connected_at"] = agent.connected_at;
        response["agents"].push_back(a);
    }
    response["count"] = agents.size();

    return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_LIST_REMOTES, response.dump());
}

ipc::Message TunnelSyscalls::handle_tunnel_config(const ipc::Message& msg) {
    try {
        json j = json::parse(msg.payload_str());

        clove::services::tunnel::TunnelConfig tc;
        tc.relay_url = j.value("relay_url", context_.config.relay_url);
        tc.machine_id = j.value("machine_id", context_.config.machine_id);
        tc.token = j.value("token", context_.config.machine_token);
        tc.reconnect_interval = j.value("reconnect_interval", 5);

        if (context_.tunnel_client.configure(tc)) {
            context_.config.relay_url = tc.relay_url;
            context_.config.machine_id = tc.machine_id;
            context_.config.machine_token = tc.token;

            json response;
            response["success"] = true;
            return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONFIG, response.dump());
        } else {
            json response;
            response["success"] = false;
            response["error"] = "Failed to configure tunnel";
            return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONFIG, response.dump());
        }

    } catch (const std::exception& e) {
        json response;
        response["success"] = false;
        response["error"] = std::string("invalid request: ") + e.what();
        return ipc::Message(msg.agent_id, ipc::SyscallOp::SYS_TUNNEL_CONFIG, response.dump());
    }
}

void TunnelSyscalls::process_tunnel_events() {
    auto events = context_.tunnel_client.poll_events();

    for (const auto& event : events) {
        switch (event.type) {
            case clove::services::tunnel::TunnelEvent::Type::SYSCALL:
                handle_tunnel_syscall(event.agent_id, event.opcode, event.payload);
                break;

            case clove::services::tunnel::TunnelEvent::Type::AGENT_CONNECTED:
                spdlog::info("Remote agent connected: {} (id={})",
                            event.agent_name, event.agent_id);
                break;

            case clove::services::tunnel::TunnelEvent::Type::AGENT_DISCONNECTED:
                spdlog::info("Remote agent disconnected: id={}", event.agent_id);
                break;

            case clove::services::tunnel::TunnelEvent::Type::DISCONNECTED:
                spdlog::warn("Tunnel disconnected from relay");
                break;

            case clove::services::tunnel::TunnelEvent::Type::RECONNECTED:
                spdlog::info("Tunnel reconnected to relay");
                break;

            case clove::services::tunnel::TunnelEvent::Type::ERROR:
                spdlog::error("Tunnel error: {}", event.error);
                break;
        }
    }
}

void TunnelSyscalls::handle_tunnel_syscall(uint32_t agent_id, uint8_t opcode,
                                          const std::vector<uint8_t>& payload) {
    ipc::Message msg;
    msg.agent_id = agent_id;
    msg.opcode = static_cast<ipc::SyscallOp>(opcode);
    msg.payload = payload;

    spdlog::debug("Processing syscall from remote agent {}: opcode=0x{:02x}",
                  agent_id, opcode);

    auto response = dispatch_(msg);

    context_.tunnel_client.send_response(
        agent_id,
        static_cast<uint8_t>(response.opcode),
        response.payload
    );
}

} // namespace clove::kernel

File: agents/python_sdk/clove_sdk/mixins/tunnel.py
"""Tunnel (remote connectivity) syscalls.

Provides relay server connection and remote agent management.
"""

from typing import Optional, TYPE_CHECKING

from ..protocol import SyscallOp
from ..models import TunnelStatus, TunnelRemotesResult, OperationResult

if TYPE_CHECKING:
    from ..transport import Transport


class TunnelMixin:
    """Mixin for tunnel/relay operations.

    Requires _transport attribute of type Transport.
    """

    _transport: 'Transport'

    def tunnel_connect(
        self,
        relay_url: str,
        machine_id: Optional[str] = None,
        token: Optional[str] = None
    ) -> TunnelStatus:
        """Connect the kernel to a relay server for remote agent access.

        Args:
            relay_url: WebSocket URL of relay server
            machine_id: Optional machine identifier
            token: Authentication token

        Returns:
            TunnelStatus with connection info
        """
        payload = {"relay_url": relay_url}
        if machine_id:
            payload["machine_id"] = machine_id
        if token:
            payload["token"] = token

        result = self._transport.call_json(SyscallOp.SYS_TUNNEL_CONNECT, payload)

        return TunnelStatus(
            success=result.get("success", False),
            connected=result.get("connected", False),
            relay_url=result.get("relay_url"),
            machine_id=result.get("machine_id"),
            latency_ms=result.get("latency_ms"),
            connected_since=result.get("connected_since"),
            error=result.get("error")
        )

    def tunnel_disconnect(self) -> OperationResult:
        """Disconnect the kernel from the relay server.

        Returns:
            OperationResult with success status
        """
        result = self._transport.call_json(SyscallOp.SYS_TUNNEL_DISCONNECT, {})

        return OperationResult(
            success=result.get("success", False),
            error=result.get("error")
        )

    def tunnel_status(self) -> TunnelStatus:
        """Get the current tunnel connection status.

        Returns:
            TunnelStatus with connection info
        """
        result = self._transport.call_json(SyscallOp.SYS_TUNNEL_STATUS, {})

        return TunnelStatus(
            success=result.get("success", False),
            connected=result.get("connected", False),
            relay_url=result.get("relay_url"),
            machine_id=result.get("machine_id"),
            latency_ms=result.get("latency_ms"),
            connected_since=result.get("connected_since"),
            error=result.get("error")
        )

    def tunnel_list_remotes(self) -> TunnelRemotesResult:
        """List remote agents currently connected through the tunnel.

        Returns:
            TunnelRemotesResult with list of remote agents
        """
        result = self._transport.call_json(SyscallOp.SYS_TUNNEL_LIST_REMOTES, {})

        return TunnelRemotesResult(
            success=result.get("success", False),
            agents=result.get("agents", []),
            count=result.get("count", 0),
            error=result.get("error")
        )

    def tunnel_config(
        self,
        relay_url: Optional[str] = None,
        machine_id: Optional[str] = None,
        token: Optional[str] = None,
        reconnect_interval: Optional[int] = None
    ) -> TunnelStatus:
        """Configure tunnel settings without connecting.

        Args:
            relay_url: WebSocket URL of relay server
            machine_id: Machine identifier
            token: Authentication token
            reconnect_interval: Reconnect interval in seconds

        Returns:
            TunnelStatus with current config
        """
        payload = {}
        if relay_url:
            payload["relay_url"] = relay_url
        if machine_id:
            payload["machine_id"] = machine_id
        if token:
            payload["token"] = token
        if reconnect_interval is not None:
            payload["reconnect_interval"] = reconnect_interval

        result = self._transport.call_json(SyscallOp.SYS_TUNNEL_CONFIG, payload)

        return TunnelStatus(
            success=result.get("success", False),
            connected=result.get("connected", False),
            relay_url=result.get("relay_url"),
            machine_id=result.get("machine_id"),
            error=result.get("error")
        )
